### PR TITLE
Deterministic Alarms and Conditions

### LINKED ISSUE #45: Feature: Deterministic testing of Alarms and Condition
> Please provide us with the following information:
> ---------------------------------------------------------------

### This issue is for a: (mark with an `x`)
```
- [ ] bug report -> please search issues before submitting
- [x] feature request
- [ ] documentation issue or request
- [ ] regression (a behavior that used to work and stopped in a new release)
```

This is a proposal to include deterministic testing for Alarms and Condition.
I have created the proposed solution and documentation in my own repository.
https://github.com/buzzfrog/MySimpleOpcAlarmServerSim

### Mention any other details that might be useful

> ---------------------------------------------------------------
> Thanks! We'll be in touch soon.

### PR DESCRIPTION
**Pull request by algattik**
## Purpose
* Enable deterministic alarms
Closes #45 

## Does this introduce a breaking change?
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test

* Test the code
- Run test suite (`DeterministicAlarmsTests`)
- Or run manually with `dotnet run --project src -- --dalm=./tests/DeterministicAlarmsTests/dalm001.json` and verify alarms and conditions in UAExpert

## What to Check
Verify that the following are valid
* Alarms and conditions occur as described in `dalm001.json` scenario

### CODE REVIEW

**Review by peterbryntesson — APPROVED** [review]
LGTM

**Review by simonjaeger — APPROVED** [review]

File: src/DeterministicAlarms/Model/SimSourceNodeState.cs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        private void ReportChanges(ConditionState alarm)
+        {
+            // report changes to node attributes.
+            alarm.ClearChangeMasks(_nodeManager.SystemContext, true);
+
+            // check if events are being monitored for the source.
+            if (this.AreEventsMonitored)
+            {
+                // create a snapshot.
+                InstanceStateSnapshot e = new InstanceStateSnapshot();
```
Comment (luiscantero) [inline src/DeterministicAlarms/Model/SimSourceNodeState.cs:260]: var

File: src/DeterministicAlarms/SimBackend/SimSourceNodeBackend.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+        public AlarmChangedEventHandler OnAlarmChanged;
+
+        public delegate void AlarmChangedEventHandler(SimAlarmStateBackend alarm);
+
+        public void CreateAlarms(List<Alarm> alarmsFromConfiguration)
+        {
+            foreach (var alarmConfiguration in alarmsFromConfiguration)
+            {
+                SimAlarmStateBackend alarmStateBackend = new SimAlarmStateBackend
```
Comment (luiscantero) [inline src/DeterministicAlarms/SimBackend/SimSourceNodeBackend.cs:21]: var

File: src/DeterministicAlarms/SimBackend/SimSourceNodeBackend.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                lock (this.Alarms)
+                {
+                    this.Alarms.Add(alarmConfiguration.Id, alarmStateBackend);
+                }
+            }
+        }
+
+        internal void Refresh()
+        {
+            List<SimAlarmStateBackend> snapshots = new List<SimAlarmStateBackend>();
```
Comment (luiscantero) [inline src/DeterministicAlarms/SimBackend/SimSourceNodeBackend.cs:42]: var

**Review by luiscantero — APPROVED** [review]

**Pull request merged**